src/StripStylesTransformer.ts

```diff
@@ -0,0 +1,150 @@
+/*
+ * Code is inspired by
+ * https://github.com/kulshekhar/ts-jest/blob/25e1c63dd3797793b0f46fa52fdee580b46f66ae/src/transformers/hoist-jest.ts
+ *
+ * 
+ * IMPLEMENTATION DETAILS:
+ * This transformer handles one concern: removing styles.
+ *
+ * The property 'styles' inside a @Component(...) Decorator argument will
+ * be modified, even if they are not used in the context of an
+ * Angular Component.
+ * 
+ * This is the required AST to trigger the transformation:
+ * 
+ * ClassDeclaration
+ *   Decorator
+ *     CallExpression
+ *       ObjectLiteralExpression
+ *         PropertyAssignment
+ *           Identifier
+ *           StringLiteral
+ */
+
+// only import types, for the rest use injected `ConfigSet.compilerModule`
+import {
+  Node,
+  SourceFile,
+  TransformationContext,
+  Transformer,
+  Visitor,
+  Identifier,
+  ClassDeclaration,
+  PropertyAssignment
+} from 'typescript';
+import { ConfigSet } from './TransformUtils';
+
+/** Angular component decorator Styles property name */
+const STYLES = 'styles';
+/** Angular component decorator name */
+const COMPONENT = 'Component';
+/** All props to be transformed inside a decorator */
+const TRANSFORM_IN_DECORATOR_PROPS = [STYLES];
+
+/**
+ * Transformer ID
+ * @internal
+ */
+export const name = 'angular-component-strip-styles';
+
+// increment this each time the code is modified
+/**
+ * Transformer Version
+ * @internal
+ */
+export const version = 1;
+
+/**
+ * The factory of hoisting transformer factory
+ * @internal
+ */
+export function factory(cs: ConfigSet) {
+  /**
+   * Our compiler (typescript, or a module with typescript-like interface)
+   */
+  const ts = cs.compilerModule;
+
+  /**
+   * Traverses the AST down inside a decorator to a styles assignment
+   * and returns a boolean indicating if it should be transformed.
+   */
+  function isInDecoratorPropertyAssignmentToTransform(node: Node): node is ClassDeclaration {
+    return getInDecoratorPropertyAssignmentsToTransform(node).length > 0;
+  }
+
+  /**
+   * Traverses the AST down inside a decorator to a styles assignment
+   * returns it in an array.
+   */
+  function getInDecoratorPropertyAssignmentsToTransform(node: Node) {
+    if (!ts.isClassDeclaration(node) || !node.decorators) {
+      return [];
+    }
+
+    return node.decorators
+      .map(dec => dec.expression)
```

Comment:
How big this array can be realistically? If it's 1000+ elements, it may make sense to squash multiple filters/maps into a single call. The best thing would be to actually measure on a cold cache :)

Reply:
**TL;DR**
It usually has only 1-4 elements, this is in my opinion the shortest and most readable version.

**Explanation**
`node.decorators` is the array that contains all decorators of a class.
Most classes anyway have only one decorator, in that case the whole function will be traversed only once, maybe even aborted, if it is a wrong decorator or the expressions inside are wrong.
In the two `flatMaps` (here `reduce`) they map to all the property assignments in the decorator, so in each decorator there are like 2-4 I would say. These get filtered out in the very end to only `styleUrls` and `templateUrl`, so the last two filters are called for each property assignment in a `@Component` decorator in the codebase.

What I personally value more is that else might have to refactor the code in some months and it is better to understand it easier and to write less code.

Finally TypeScript has better type support when using a single statement in a filter (e. g. `.filter(ts.isCallExpression)`), it tells me without cast that the argument in the next `map`/`filter` is of the type `CallExpression`, due to the return type of the function `is CallExpression`. The same does not happen when chaining with `&&`, so otherwise there will be more casting and assignments to casted variables.

Hope this explains my view, but I'm open to make changes.

**Benchmarking**
I have also no data to support this, but I believe the performance-wise expensive part of the transformers are when things are modified. I think this is much more expensive than iterating over some parts of the asts.

With this new implementation we modify less `styles` assignments and so this should be faster. But I am also sure that the tasks done by `ts-jest`, like building the initial ast and transpiling the code afterwards are much more heavier than these small transformations. Probably the optimizations of the v8-engine weigh much more than code changes in the transformers.

I am critical about benchmarking in general, I think to get a meaningful result we should test a bigger codebase with more components. Maybe there is a big open source Angular project out there?

In the CI-test-process we could clone a repo and test it with disabled jest-cache.

Reply:
That makes perfect sense, thanks for the detailed writeup :)